indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java

```diff
@@ -576,16 +577,105 @@ private static void createNewZipEntry(ZipOutputStream out, File file) throws IOE
     out.putNextEntry(new ZipEntry(file.getName()));
   }
 
-  public static Path makeSegmentOutputPath(
-      Path basePath,
-      FileSystem fileSystem,
-      DataSegment segment
+  public static boolean isHdfs(FileSystem fs)
+  {
+    return "hdfs".equals(fs.getScheme()) || "viewfs".equals(fs.getScheme());
+  }
+
+  public static Path makeIndexZipPath(
+      final Path basePath,
+      final FileSystem fs,
+      final DataSegment segmentTemplate
   )
   {
-    String segmentDir = "hdfs".equals(fileSystem.getScheme()) || "viewfs".equals(fileSystem.getScheme())
-                        ? DataSegmentPusherUtil.getHdfsStorageDir(segment)
-                        : DataSegmentPusherUtil.getStorageDir(segment);
-    return new Path(prependFSIfNullScheme(fileSystem, basePath), String.format("./%s", segmentDir));
+    final Path finalIndexZipPath;
+    final String segmentDir;
+    if (isHdfs(fs)) {
+      segmentDir = DataSegmentPusherUtil.getHdfsStorageDir(segmentTemplate);
+      finalIndexZipPath = new Path(
+          prependFSIfNullScheme(fs, basePath),
+          String.format(
+              "./%s/%d_index.zip",
+              segmentDir,
+              segmentTemplate.getShardSpec().getPartitionNum()
+          )
+      );
+    } else {
+      segmentDir = DataSegmentPusherUtil.getStorageDir(segmentTemplate);
+      finalIndexZipPath = new Path(
+          prependFSIfNullScheme(fs, basePath),
+          String.format(
+              "./%s/index.zip",
+              segmentDir
+          )
+      );
+    }
+    return finalIndexZipPath;
+  }
+
+  public static Path makeDescriptorPath(
```

Comment:
If you want, this and `makeIndexZipPath` could share a lot of logic in a common helper method that just takes the base file name ("index.zip" vs "descriptor.json")